### ISSUE TITLE
Mount and unmount with MutationObserver

### ISSUE THREAD

**Issue by hexpl0it** [wxt-dev/wxt#537]
For my extension I need to mount my app immediately after a precise div. However, this div is not immediately available when the page is loaded.

To do this I used MutationObserver: 

```javascript
import ReactDOM from "react-dom/client";
import Toolbar from "./components/Toolbar";
import { useEffect } from "react";

function waitForElm(selector) {
  return new Promise((resolve) => {
    if (document.querySelector(selector)) {
      return resolve(document.querySelector(selector));
    }
    const observer = new MutationObserver((mutations) => {
      if (document.querySelector(selector)) {
        observer.disconnect();
        resolve(document.querySelector(selector));
      }
    });
    observer.observe(document.body, {
      childList: true,
      subtree: true,
    });
  });
}

export default defineContentScript({
  main(ctx) {
    const ui = createIntegratedUi(ctx, {
      position: "inline",
      onMount: async (container) => {
        const elm = await waitForElm('[class^="ToolbarContainer__StyledHeader"]');
        const rootNode = document.createElement("div");
        elm.insertAdjacentElement("afterend", rootNode);
        const root = ReactDOM.createRoot(rootNode);
        root.render(<Toolbar />);
        return root;
      },
      onRemove: (root) => {
        root.unmount();
      },
    });

    ui.mount()
  },
});
``` 

However, this div can be removed following certain actions on the page. I would need to reassemble my component as soon as this div reappears. How can I do this?

**Comment by aklinker1** [comment]
You'll want to use the awaited element as the `anchor` tag.

Something like this would work. (I typed this up in github comments, so it probably isn't valid JS, and I haven't tested it).

```ts
let anchor;
const ui  = createIntegratedUi(ctx, {
  position: "inline",
  anchor: () => anchor,
  onMount: async (container) => {
    const root = ReactDOM.createRoot(container);
    root.render(<Toolbar />);
    return root;
  },
  onRemove: (root) => {
    root.unmount();
  },
});
watchDomChanges(ctx, '[class^="ToolbarContainer__StyledHeader"]', {
  onAdd: (newAnchor) => {
    anchor = newAnchor;
    ui.mount();
  },
  onRemove: () => {
    ui.remove();
  },
});
```

```js
function watchDomChanges(ctx, selector, callbacks) {
  const addMountListener = () => {
    const observer = new MutationObserver((mutations) => {
      const el = document.querySelector(selector)
      if (el) {
        observer.disconnect();
        unwatchCtx();
        callbacks.onAdd(el);
        addRemoveListener(el);
      }
    });
    const unwatchCtx = ctx.onInvalidated(observer.disconnect);
    observer.observe(document.body, {
      childList: true,
      subtree: true,
    });
  };
  const addRemoveListener = (el) => {
    const observer = new MutationObserver((mutations) => {
      if (mutations[0].removedNodes.includes(el)) {
        observer.disconnect();
        unwatchCtx();
        callbacks.onRemoved();
        addMountListener();
      }
    });
    const unwatchCtx = ctx.onInvalidated(observer.disconnect);
    observer.observe(el);
  };
  
  const initialEl = document.querySelector(selector);
  if (initialEl) {
    callbacks.onAdd(initialEl);
    addRemoveListener(initialEl);
  } else {
    addMountListener();
  }
}
```

**Comment by aklinker1** [comment]
@hexpl0it If you get something that works, I'd like to add auto-mounting/unmounting to WXT.

**Comment by tesths** [comment]
Thank you very much for this issue. I have recently encountered a similar problem.
I want to inject content script above a specific div on the website after it has loaded.
I have tried the following methods, which are effective. Thanks for @aklinker1 .

> You'll want to use the awaited element as the `anchor` tag, and use the `append` option to append your UI using `insertAdjacentElement`.
> 
> Something like this would work. I typed this up in github comments, so it probably isn't valid JS, and I haven't tested it. But hopefully it gives you a better idea of what you're looking for.
> 
> ```ts
> let anchor;
> const ui  = createIntegratedUi(ctx, {
>   position: "inline",
>   anchor: () => anchor,
>   append: (anchor, root) => anchor.insertAdjacentElement("afterend", root),
>   onMount: async (container) => {
>     const root = ReactDOM.createRoot(container);
>     root.render(<Toolbar />);
>     return root;
>   },
>   onRemove: (root) => {
>     root.unmount();
>   },
> });
> watchDomChanges(ctx, '[class^="ToolbarContainer__StyledHeader"]', {
>   onAdd: (newAnchor) => {
>     anchor = newAnchor;
>     ui.mount();
>   },
>   onRemove: () => {
>     ui.remove();
>   },
> });
> ```
> 
> ```js
> function watchDomChanges(ctx, selector, callbacks) {
>   const addMountListener = () => {
>     const observer = new MutationObserver((mutations) => {
>       const el = document.querySelector(selector)
>       if (el) {
>         observer.disconnect();
>         unwatchCtx();
>         callbacks.onAdd(el);
>         addRemoveListener(el);
>       }
>     });
>     const unwatchCtx = ctx.onInvalidated(observer.disconnect);
>     observer.observe(document.body, {
>       childList: true,
>       subtree: true,
>     });
>   };
>   const addRemoveListener = (el) => {
>     const observer = new MutationObserver((mutations) => {
>       if (mutations[0].removedNodes.includes(el)) {
>         observer.disconnect();
>         unwatchCtx();
>         callbacks.onRemoved();
>         addMountListener();
>       }
>     });
>     const unwatchCtx = ctx.onInvalidated(observer.disconnect);
>     observer.observe(el);
>   };
>   
>   const initialEl = document.querySelector(selector);
>   if (initialEl) {
>     callbacks.onAdd(initialEl);
>     addRemoveListener(initialEl);
>   } else {
>     addMountListener();
>   }
> }
> ```

**Comment by aklinker1** [comment]
Think I'm gonna work on this soon. Here's what I'm thinking the API will look like:

```ts
const ui = createXyzUi({
  anchor: "#some-anchor",
})

ui.autoMount();
```

Here's the types I expect:

```ts
type StopAutoMount = () => void;

interface UI {
  autoMount(options?: { once?: boolean }): StopAutoMount;
}
```

**Comment by 1natsu172** [comment]
I've been thinking about this Issue in the back of my mind for a long time. We get a lot of questions about dynamic UI mounts. 

I just wonder what and how WXT should support. I have a feeling that `autoMount` will have a lot of responsibility since the timing and conditions are different for each application.

BTW: Actually, I have a [personal library](https://github.com/1natsu172/wait-element) for dynamic mounting, lol.

**Comment by aklinker1** [comment]
> I have a feeling that `autoMount` will have a lot of responsibility since the timing and conditions are different for each.

Hmm, I recommended this function to addressing a single use case: mounting a UI inside a dynamic element that gets added and removed.

If someone wants a custom implementation, like listening to focus events, or adding timeouts, they should write those themselves using `ui.mount` and `ui.unmount`. that's why we provide those APIs.

> BTW: Actually, I have a personal library for dynamic mounting, lol.

Nice! We can probably use it, does it support listening for when the element is removed from the DOM?

**Comment by 1natsu172** [comment]
> If someone wants a custom implementation, like listening to focus events, or adding timeouts, they should write those themselves using ui.mount and ui.unmount. that's why we provide those APIs.

If you are set in your mind on that policy, it's ok 👍 . Writing in the document that it is a single use case would avoid confusion users.

> does it support listening for when the element is removed from the DOM?

Supported by [`detector` option](https://github.com/1natsu172/wait-element#custom-detect-condition). However, this library doesn't manage DOM state. This is an async wrapper for pure `mutationObserver` and `querySelector`. So if you want accuracy based on DOM state, will need a separate DOM state manager.

Basically, it goes like this. (recently, major bump to v4, so if there any bugs it might not work properly 😅 ）

```ts
// waiting anchor
const anchor = await waitElement(targetAnchor);
if (anchor) {
  ui.mount();
}
// waiting remove
const nowAnchor = await waitElement(targetAnchor, { detector: isNotExist });
if (nowAnchor === null) {
  ui.unmount()
}
```

**Comment by Ya-hui** [comment]
Hello, I am using your package, but the background page still shows an error message:
The content script "ali-product" crashed on startup! Error: Failed to mount content script UI: could not find anchor element

**Comment by aklinker1** [comment]
I am onboard with adding functionality to WXT.

**Comment by 1natsu172** [comment]
Yeah, I’m willing to take this. I recently made my library support xpath for this issue :). I'm trying to write the test code first……but recently I've been so busy that I haven't had time. Need more time.

**Comment by sgcullen** [comment]
Interesting discussion - I'm doing similar stuff - following :-)

**Comment by aklinker1** [comment]
> About once option API, is this “only one mount” or “one mount and one remove”? Which did you expect?

I would expect it to be mounted and removed once.

> And I think about the `StopAutoMount`, should be stopped when `ui.remove` is called, is there any discrepancy?

Yes, I agree with you on that. Calling `ui.remove()` should stop the auto-mount watcher as well.

**Comment by 1natsu172** [comment]
Okay, I'll go ahead with those specs for both!

**Comment by aklinker1** [comment]
Release in [`v0.19.20`](https://github.com/wxt-dev/wxt/releases/tag/wxt-v0.19.20): https://wxt.dev/guide/essentials/content-scripts.html#mounting-ui-to-dynamic-element
